File: laravel-project/tests/Feature/Api/OrderTest.php
<?php

namespace Tests\Feature\Api;

use App\Models\Order;
use App\Models\Product;
use App\Models\User;
use Tests\TestCase;
use Illuminate\Foundation\Testing\RefreshDatabase;

class OrderTest extends TestCase
{
    use RefreshDatabase;

    /** @test */
    public function authenticated_user_can_list_all_orders()
    {
        $user = User::factory()->create();
        Order::factory()->count(3)->create();

        $response = $this->actingAs($user, 'sanctum')
            ->getJson('/api/orders');

        $response->assertStatus(200)
            ->assertJsonCount(3);
    }

    /** @test */
    public function unauthenticated_user_cannot_list_orders()
    {
        Order::factory()->count(2)->create();

        $response = $this->getJson('/api/orders');

        $response->assertStatus(401);
    }

    /** @test */
    public function authenticated_user_can_view_single_order()
    {
        $user = User::factory()->create();
        $order = Order::factory()->create([
            'user_id' => $user->id,
            'status' => 'pending',
            'total_amount' => 100.00,
        ]);

        $response = $this->actingAs($user, 'sanctum')
            ->getJson("/api/orders/{$order->id}");

        $response->assertStatus(200)
            ->assertJson([
                'id' => $order->id,
                'status' => 'pending',
                'total_amount' => 100.00,
            ]);
    }

    /** @test */
    public function authenticated_user_can_create_order()
    {
        $user = User::factory()->create();
        $product1 = Product::factory()->create(['price' => 50.00]);
        $product2 = Product::factory()->create(['price' => 30.00]);

        $orderData = [
            'items' => [
                [
                    'product_id' => $product1->id,
                    'quantity' => 2,
                ],
                [
                    'product_id' => $product2->id,
                    'quantity' => 1,
                ],
            ],
        ];

        $response = $this->actingAs($user, 'sanctum')
            ->postJson('/api/orders', $orderData);

        $response->assertStatus(201)
            ->assertJsonStructure([
                'id',
                'user_id',
                'status',
                'total_amount',
            ]);

        $this->assertDatabaseHas('orders', [
            'user_id' => $user->id,
            'status' => 'pending',
            'total_amount' => 130.00, // (50 * 2) + (30 * 1)
        ]);
    }

    /** @test */
    public function order_creation_requires_items()
    {
        $user = User::factory()->create();

        $response = $this->actingAs($user, 'sanctum')
            ->postJson('/api/orders', []);

        $response->assertStatus(422)
            ->assertJsonValidationErrors(['items']);
    }

    /** @test */
    public function order_items_require_product_id_and_quantity()
    {
        $user = User::factory()->create();
        $orderData = [
            'items' => [
                [
                    'product_id' => 1,
                    // missing quantity
                ],
            ],
        ];

        $response = $this->actingAs($user, 'sanctum')
            ->postJson('/api/orders', $orderData);

        $response->assertStatus(422);
    }

    /** @test */
    public function authenticated_user_can_update_order_status()
    {
        $user = User::factory()->create();
        $order = Order::factory()->create([
            'user_id' => $user->id,
            'status' => 'pending',
        ]);

        $response = $this->actingAs($user, 'sanctum')
            ->putJson("/api/orders/{$order->id}", [
                'status' => 'completed',
            ]);

        $response->assertStatus(200)
            ->assertJson([
                'status' => 'completed',
            ]);

        $this->assertDatabaseHas('orders', [
            'id' => $order->id,
            'status' => 'completed',
        ]);
    }

    /** @test */
    public function order_calculates_total_correctly_with_multiple_items()
    {
        $user = User::factory()->create();
        $product1 = Product::factory()->create(['price' => 25.00]);
        $product2 = Product::factory()->create(['price' => 15.50]);
        $product3 = Product::factory()->create(['price' => 10.00]);

        $orderData = [
            'items' => [
                ['product_id' => $product1->id, 'quantity' => 3], // 75.00
                ['product_id' => $product2->id, 'quantity' => 2], // 31.00
                ['product_id' => $product3->id, 'quantity' => 5], // 50.00
            ],
        ];

        $response = $this->actingAs($user, 'sanctum')
            ->postJson('/api/orders', $orderData);

        $response->assertStatus(201);

        $this->assertDatabaseHas('orders', [
            'user_id' => $user->id,
            'total_amount' => 156.00, // 75 + 31 + 50
        ]);
    }

    /** @test */
    public function it_returns_404_for_nonexistent_order()
    {
        $user = User::factory()->create();

        $response = $this->actingAs($user, 'sanctum')
            ->getJson('/api/orders/99999');

        $response->assertStatus(404);
    }

    /** @test */
    public function order_with_items_creates_order_items_records()
    {
        $user = User::factory()->create();
        $product = Product::factory()->create(['price' => 100.00]);

        $orderData = [
            'items' => [
                [
                    'product_id' => $product->id,
                    'quantity' => 2,
                ],
            ],
        ];

        $response = $this->actingAs($user, 'sanctum')
            ->postJson('/api/orders', $orderData);

        $response->assertStatus(201);

        $orderId = $response->json('id');

        $this->assertDatabaseHas('order_items', [
            'order_id' => $orderId,
            'product_id' => $product->id,
            'quantity' => 2,
            'price' => 100.00,
        ]);
    }

    /** @test */
    public function get_order_stats_returns_orders_by_status()
    {
        $user = User::factory()->create();
        Order::factory()->pending()->count(3)->create();
        Order::factory()->completed()->count(2)->create();

        $response = $this->actingAs($user, 'sanctum')
            ->getJson('/api/orders/stats?status=pending');

        $response->assertStatus(200)
            ->assertJsonCount(3);
    }

    /** @test */
    public function order_includes_user_relationship()
    {
        $user = User::factory()->create(['name' => 'John Doe']);
        $order = Order::factory()->create(['user_id' => $user->id]);

        $response = $this->actingAs($user, 'sanctum')
            ->getJson('/api/orders');

        $response->assertStatus(200)
            ->assertJsonFragment([
                'name' => 'John Doe',
            ]);
    }
}

File: flutter-project/test/models/order_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:qa_assessment_app/models/order.dart';

void main() {
  group('OrderItem Model', () {
    test('should create OrderItem from JSON', () {
      final json = {
        'id': 1,
        'product_id': 10,
        'quantity': 2,
        'price': 99.99,
      };

      final orderItem = OrderItem.fromJson(json);

      expect(orderItem.id, 1);
      expect(orderItem.productId, 10);
      expect(orderItem.quantity, 2);
      expect(orderItem.price, 99.99);
    });

    test('should handle null price in OrderItem', () {
      final json = {
        'id': 2,
        'product_id': 20,
        'quantity': 1,
        'price': null,
      };

      final orderItem = OrderItem.fromJson(json);

      expect(orderItem.price, 0.0);
    });

    test('should convert OrderItem to JSON', () {
      final orderItem = OrderItem(
        id: 3,
        productId: 30,
        quantity: 5,
        price: 25.50,
      );

      final json = orderItem.toJson();

      expect(json['id'], 3);
      expect(json['product_id'], 30);
      expect(json['quantity'], 5);
      expect(json['price'], 25.50);
    });
  });

  group('Order Model', () {
    test('should create Order from JSON without items', () {
      final json = {
        'id': 100,
        'user_id': 1,
        'status': 'pending',
        'total_amount': 199.99,
      };

      final order = Order.fromJson(json);

      expect(order.id, 100);
      expect(order.userId, 1);
      expect(order.status, 'pending');
      expect(order.totalAmount, 199.99);
      expect(order.items, isNull);
    });

    test('should create Order from JSON with items', () {
      final json = {
        'id': 101,
        'user_id': 2,
        'status': 'completed',
        'total_amount': 299.98,
        'items': [
          {
            'id': 1,
            'product_id': 10,
            'quantity': 2,
            'price': 99.99,
          },
          {
            'id': 2,
            'product_id': 20,
            'quantity': 1,
            'price': 100.0,
          },
        ],
      };

      final order = Order.fromJson(json);

      expect(order.id, 101);
      expect(order.userId, 2);
      expect(order.status, 'completed');
      expect(order.totalAmount, 299.98);
      expect(order.items, isNotNull);
      expect(order.items!.length, 2);
      expect(order.items![0].productId, 10);
      expect(order.items![1].productId, 20);
    });

    test('should handle null total_amount', () {
      final json = {
        'id': 102,
        'user_id': 3,
        'status': 'cancelled',
        'total_amount': null,
      };

      final order = Order.fromJson(json);

      expect(order.totalAmount, 0.0);
    });

    test('should convert Order to JSON without items', () {
      final order = Order(
        id: 103,
        userId: 4,
        status: 'shipped',
        totalAmount: 499.99,
      );

      final json = order.toJson();

      expect(json['id'], 103);
      expect(json['user_id'], 4);
      expect(json['status'], 'shipped');
      expect(json['total_amount'], 499.99);
      expect(json['items'], isNull);
    });

    test('should convert Order to JSON with items', () {
      final order = Order(
        id: 104,
        userId: 5,
        status: 'delivered',
        totalAmount: 599.99,
        items: [
          OrderItem(id: 1, productId: 10, quantity: 3, price: 199.99),
          OrderItem(id: 2, productId: 20, quantity: 1, price: 200.0),
        ],
      );

      final json = order.toJson();

      expect(json['id'], 104);
      expect(json['user_id'], 5);
      expect(json['status'], 'delivered');
      expect(json['total_amount'], 599.99);
      expect(json['items'], isNotNull);
      expect(json['items'].length, 2);
    });
  });
}
